includes/class-wc-stripe-apple-pay-registration.php

```diff
@@ -90,6 +99,62 @@ private function get_secret_key() {
 		return $testmode ? $this->get_option( 'test_secret_key' ) : $this->get_option( 'secret_key' );
 	}
 
+	/**
+	 * Trigger Apple Pay registration upon domain name change.
+	 *
+	 * @since 4.9.0
+	 */
+	public function verify_domain_on_domain_name_change() {
+		if ( $this->domain_name !== $this->get_option( 'apple_pay_verified_domain' ) ) {
+			$this->verify_domain_if_configured();
+		}
+	}
+
+	/**
+	 * Updates the Apple Pay domain association file.
+	 * Reports failure only if file isn't already being served properly.
+	 *
+	 * @since 4.9.0
+	 */
+	public function update_domain_association_file() {
+		$path     = untrailingslashit( ABSPATH );
+		$dir      = '.well-known';
+		$file     = 'apple-developer-merchantid-domain-association';
+		$fullpath = $path . '/' . $dir . '/' . $file;
+
+		$existing_contents = @file_get_contents( $fullpath ); // @codingStandardsIgnoreLine
+		$new_contents      = @file_get_contents( WC_STRIPE_PLUGIN_PATH . '/' . $file ); // @codingStandardsIgnoreLine
+		if ( $existing_contents === $new_contents ) {
+			return;
+		}
+
+		$error = null;
+		if ( ! file_exists( $path . '/' . $dir ) ) {
+			if ( ! @mkdir( $path . '/' . $dir, 0755 ) ) { // @codingStandardsIgnoreLine
+				$error = __( 'Unable to create domain association folder to domain root.', 'woocommerce-payments' );
+			}
+		}
+
+		if ( ! @copy( WC_STRIPE_PLUGIN_PATH . '/' . $file, $fullpath ) ) { // @codingStandardsIgnoreLine
+			$error = __( 'Unable to copy domain association file to domain root.', 'woocommerce-payments' );
+		}
```

Comment:
This whole `update_domain_association_file` function was restored from a previous implementation and just adapted here as a fallback verification method.

In order to separate these functions I need to do something about the shared variables. I'll suggest something and then you can review.

Reply:
@reykjalin, I agree this function is not so easy to read. I started refactoring it here, but it started looking worse, because we have to pass those shared variables or re-declare them ($dir, $file, $path, $fullpath) and the first verification at the top...

```php
$existing_contents = @file_get_contents( $fullpath ); // @codingStandardsIgnoreLine
$new_contents      = @file_get_contents( WC_STRIPE_PLUGIN_PATH . '/' . $this->domain_association_ile ); // @codingStandardsIgnoreLine
if ( $existing_contents === $new_contents ) {
    return;
}
```

... ideally would be split too, so we'd have 4 different functions in total, thus increasing the overall complexity of this class.

After some thought, I'd suggest keeping it as is, unless you feel strongly about it and/or have a suggestion on how to make it cleaner with the shared variables in mind  🙂

Reply:
FYI the original function can be found [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1357/files#diff-b3ce8b937d7eea48160e4d6e7bd28e652b0b4834af86ac2b60bdb748aac0eedfL181). Now we're adding it again as a fallback because the rewrite rule isn't always reliable.

Reply:
I'm personally fine with creating a private helper function like `get_well_known_directory_path()`:

```php
function get_well_known_directory_path() {
	return untrailingslashit( ABSPATH ) . '/.well-known';
}
```

We can then use that helper function in `update_domain_association_file()`:

```php
function update_domain_association_file() {
	$well_known_dir = $this->get_well_known_directory_path();

	// This might even be added to a `get_local_domain_association_path()` helper function.
	$file = 'apple-developer-merchantid-domain-association';
	$fullpath = $well_known_dir . '/' . $file;

	if ( ! file_exists( $well_known_dir ) ) {
		if ( ! @mkdir( $well_known_dir, 0755 ) ) { // @codingStandardsIgnoreLine
			return __( 'Unable to create domain association folder to domain root.', 'woocommerce-payments' );
		}
	}

	if ( ! @copy( WC_STRIPE_PLUGIN_PATH . '/' . $file, $fullpath ) ) { // @codingStandardsIgnoreLine
		return __( 'Unable to copy domain association file to domain root.', 'woocommerce-payments' );
	}
}
```

Sure, it adds a bit more noise in the form of functions, but it will make the actual business logic easier to read, I think.

My overall worry here is that the work being done in the current function is difficult to parse; I want to make that as easy to follow as possible. If that means hiding the complexities in private functions that add some noise then so be it 🙂